java/arcs/android/storage/service/BindingContext.kt

```diff
@@ -190,7 +217,47 @@ class BindingContext(
         }
     }
 
+    /**
+     * Increment an active message count and send it to the active messages count flow.
+     *
+     * Calls to [idle] will wait for this value to reach 0 before checking store activity.
+     *
+     * Avoid using this for calls to [idle] itself, so as to not increment the activity counter.
+     */
+    private fun launchNonIdleAction(action: suspend () -> Unit) {
+        activeMessagesCountFlow.increment()
+        // TODO(b/168724138) - Switch to a channel/flow queue-based approach.
+        actionQueue
+            .launch(action)
+            .invokeOnCompletion { activeMessagesCountFlow.decrement() }
+    }
+
     companion object {
         private val nextId = atomic(0)
     }
 }
+
+@ExperimentalCoroutinesApi
+/**
+ * A simple helper class to provide serialized operation execution on the provided
+ * [CoroutineScope].
+ *
+ * TODO(b/168724138) - Switch to a channel/flow queue-based approach and remove this.
+ */
+class MutexOperationQueue(
```

Comment:
I feel like this is useful enough to be in arcs/core/util.

Note that this could also be achieved without a Mutex by using a Channel + Flow.

Reply:
I didn't make this a general util for exactly the reason you mention.

See the comment immediately above: `TODO(b/168724138) - Switch to a channel/flow queue-based approach and remove this.`

Reply:
Moving this to util is a great idea. We should extract all of these flow-based patterns into reusable utils!